pkg/fleetautoscalers/controller_test.go

```diff
@@ -695,6 +700,88 @@ func TestControllerUpdateStatusUnableToScale(t *testing.T) {
 	})
 }
 
+func TestControllerSyncFleetAutoscalerWithCustomSyncInterval(t *testing.T) {
+	t.Run("create fas thread", func(t *testing.T) {
+		t.Parallel()
+		c, m := newFakeController()
+		fas, f := defaultFixtures()
+		fasKey := fas.Namespace + "/" + fas.Name
+		fas.Spec.Sync.FixedInterval.Seconds = 10
+		var latestUpdatedTime time.Time
+
+		m.AgonesClient.AddReactor("list", "fleetautoscalers", func(action k8stesting.Action) (bool, runtime.Object, error) {
+			return true, &autoscalingv1.FleetAutoscalerList{Items: []autoscalingv1.FleetAutoscaler{*fas}}, nil
+		})
+
+		m.AgonesClient.AddReactor("update", "fleetautoscalers", func(action k8stesting.Action) (bool, runtime.Object, error) {
+			ca := action.(k8stesting.UpdateAction)
+			fas := ca.GetObject().(*autoscalingv1.FleetAutoscaler)
+			if !latestUpdatedTime.IsZero() {
+				assert.Equal(t, fas.Spec.Sync.FixedInterval.Seconds, int32(time.Since(latestUpdatedTime).Seconds()))
+			}
+			latestUpdatedTime = time.Now()
+			return true, fas, nil
+		})
+
+		m.AgonesClient.AddReactor("list", "fleets", func(action k8stesting.Action) (bool, runtime.Object, error) {
+			return true, &agonesv1.FleetList{Items: []agonesv1.Fleet{*f}}, nil
+		})
+
+		ctx, cancel := agtesting.StartInformers(m, c.fleetSynced, c.fleetAutoscalerSynced)
+		defer cancel()
+
+		err := c.syncFleetAutoscalerWithCustomSyncInterval(ctx, fasKey)
+		assert.Nil(t, err)
+		assert.Contains(t, c.fasThreads, fasKey)
+		// wait one sync interval, the fas update function should be called twice
+		time.Sleep(time.Duration(fas.Spec.Sync.FixedInterval.Seconds+1) * time.Second)
```

Comment:
Two thoughts:

1. Shouldn't there be something after the sleep to confirm that the function has been run twice? (maybe a counter in the update mock function?)
2. Mentioned [here](https://github.com/googleforgames/agones/pull/2171/files#r673558093), let's use [util/clock](https://pkg.go.dev/k8s.io/apimachinery/pkg/util/clock) - as then we don't need to sleep, we can instead roll time forward without have to wait for real time to pass (which also provides less chance for flakey tests).

Reply:
I tried to add something like a counter in update mock function. It works when I run the single test. However it failed when I run `make test-go` due to race condition. So I have to check the interval inside the update mock function.

For your second suggestion, do you mean use the `clock.FakeClock` to skip waiting for real time? I tried but it looks like the update function will be only called once even I wait for a longer time. If I use `clock.RealClock` it looks like no difference with `time`. Is there other way to leverage the `util/clock`?

Reply:
> I tried to add something like a counter in update mock function. It works when I run the single test. However it failed when I run make test-go due to race condition. So I have to check the interval inside the update mock function.

Wrap the counter access and mutation in a mutex, that'll solve that problem 👍🏻 

> For your second suggestion, do you mean use the clock.FakeClock to skip waiting for real time?

Yep, exactly. You will need to use an instance of the `Clock` interface in the controller, and make sure to grab the ticker from it's `NewTicker` functionality. You can no longer access the `time` package directly, everything need to run through the `Clock` instance - then in the test you can switch out the `Clock` implementation with a `FakeClock` and control the flow of perceived time by the system that way.

Reply:
Thanks for your guide! I updated the controller structure and related tests. Please take a look.